config/initializers/time_with_zone.rb

```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+# Ensure TimeWithZone are represented as RFC3339 DateTime with 3 digits of
+# precision for the fractional part of the second.
+class ActiveSupport::TimeWithZone
+  def as_json(_options = {})
```

Comment:
I know some are still pending and this is (maybe?) still somewhat WIP -- but I'm curious, are there specs here which fail when this is not present?

It's a bit of a maze to navigate because we don't always have the same starting classes (I think that AR model timestamp columns do produce `ActiveSupport::TimeWithZone`, but there may be other things using `Date`, `DateTime`, etc - or we may be overriding/casting values in place).

I mostly want to understand what this initializer is doing. When I look at the default implementation - https://github.com/rails/rails/blob/v7.2.2/activesupport/lib/active_support/time_with_zone.rb#L154-L172 - the behavior is contingent on a setting for standard json format which defaults to true and we dont override, and then also on the configured time precision, which defaults to 3.

Separately, within that class - https://github.com/rails/rails/blob/v7.2.2/activesupport/lib/active_support/time_with_zone.rb#L144-L152 - the `iso8601` and the `rfc3339` methods are both aliases for `xmlschema` which takes that precision argument and calls out to formatting methods to actually make the string.

So I guess, upon my naive reading here, it seems like the thing which the monkeypatch is doing ... is already the default behavior? Is there another override somewhere in the AMS layer or elsewhere that we are re-overriding here?

As a small example...

```
mastodon(test)> a = Time.zone.now
=> Fri, 06 Dec 2024 21:02:42.077544000 UTC +00:00
mastodon(test)> a.iso8601
=> "2024-12-06T21:02:42Z"
mastodon(test)> a.rfc3339
=> "2024-12-06T21:02:42Z"
mastodon(test)> a.as_json
=> "2024-12-06T21:02:42.077Z"
mastodon(test)> a.class
=> ActiveSupport::TimeWithZone
```

then converted...

```
mastodon(test)> b = a.to_datetime
=> Fri, 06 Dec 2024 21:02:42 +0000
mastodon(test)> b.iso8601
=> "2024-12-06T21:02:42+00:00"
mastodon(test)> b.rfc3339
=> "2024-12-06T21:02:42+00:00"
mastodon(test)> b.as_json
=> "2024-12-06T21:02:42.077+00:00"
```

So there is a difference there between the `AS::TimeWithZone` and `DateTime` json serializations ... but the default behavior of AS:TWZ seems not changed by this config?

Reply:
I think you're correct that setting it here does not change the default behaviour.

However, setting it specifically in `time_with_zone.rb` does a few things:

1. It guards against any future changes to the Rails default.
2. It guards against any code that changes the value of `.time_precision`.
3. It clearly documents intentions.

I'll extend the comment to note that this is because of an external API commitment.

Reply:
Also, to re-iterate, reading any code/comments that talk about "ISO 8601 format" should set off alarm bells that the author was being too vague and the code should be treated very carefully.

RFC3339 is a concrete, actually useful definition.

XMLSchema came after (as far as I can tell), and re-used most of the RFC3339 definition without crediting it (https://www.w3.org/TR/xmlschema-2/#dateTime). It's slightly stricter than RFC3339 (e.g., `T` as a separator rather than `T` or `t`).

Reply:
Thanks for confirming I wasn't missing something!

I guess if we're worried about the framework defaults changing and trying to protect against that, my choices in rough order of preference would be:

- Make sure the serializer specs would start failing if any of the things we care about in the json date format break during a dependency update
- If for whatever reason we can't full capture that in those specs, or we're worried about a careless developer just blindly changing the specs to "fix" the failures -- add an explicit (non serializer) spec somewhere that has a very verbose setup/description/comments/etc to make clear why we care and is purely about generic json date format serialization, and do an exact string comparison to a known date value (instead of generic "does this parse" check)
- If we need to add a configuration setting to enable either of those prior items (ie, set the precision value, set the format style, whatever), do that in an initializer
- Only as a last resort fall back to monkey patch if we can't be confident with any of the prior

Reply:
I still think this is best left out of this PR.